### ISSUE TITLE
Strange issues with Modals and Loading.dismiss() in 2.0.0-beta.10

### ISSUE THREAD

**Issue by dkobia** [driftyco/ionic#7197]
Hopefully I can describe this in a way that makes sense.

* I have a page that loads some data and fires off a loading indicator re: `this.nav.present(loading)`
* Once the data is done loading, we dismiss the loading component via `loading.dismiss();`
* On the loaded view, I tap to open a Modal

Problem:
* The Modal closes itself suddenly
* Using the console.log, I see that `loading.dismiss();` is fired again (strange), and the new dismiss() inadvertently closes the Modal.

I have tried waiting until the second `loading.dismiss();` is fired before opening the Modal again and no issues. Therefore are two problems here:

* Why is `loading.dismiss();` fired twice?? This started happening in 2.0.0-beta.10
* Why would a `loading.dismiss();` close a Modal?

**Comment by jgw96** [comment]
Hello, thanks for opening an issue with us! Would you be able to provide a plunker that demonstrates this issue? Thanks for using Ionic!

**Comment by flash716** [comment]
I'm also seeing this behavior. It only happens when a duration is set on the loader. It appears that the second time onDismiss is called is when the timeout is hit.

The below plnkr example should pop 2 browser alerts, one at 5 seconds when loading.dismiss() is called and the second at 10 seconds when the duration is reached.

http://plnkr.co/edit/TZNsTUTUpkQ6HBj9V6tw?p=preview

**Comment by dkobia** [comment]
@flash716 thanks for the plnkr. I didn't have time to put one together.

**Comment by chrskrchr** [comment]
Not sure if it's the same issue or not, but I was running into something similar with pushing/popping views and presenting/dismissing loading overlays in parallel that was caused by the race condition described here: 

> "Nav Transition Promises" - http://ionicframework.com/docs/v2/api/components/nav/NavController/
Navigation transitions are asynchronous, meaning they take a few moments to finish, and the duration of a transition could be any number. In most cases the async nature of a transition doesn't cause any problems and the nav controller is pretty good about handling which transition was the most recent when multiple transitions have been kicked off. However, when an app begins firing off many transitions, on the same stack at roughly the same time, the nav controller can start to get lost as to which transition should be finishing, and which transitions should not be animated.

If you look at the code for pushing views and presenting/dismissing loading overlays, you'll see that the views all end up on the same view stack and, as described in the docs, things can get out of whack if you kick off multiple transitions that modify that stack.  

I ultimately solved the issue by creating a transition scheduler that queues transitions in the order in which they were initiated, and ensures that only one transition is ever executing at a given point in time.  This isn't exactly ideal because I'm now forcing the user to wait for these transitions to complete before the user can continue using the app, but it's better than ending up with a corrupt view stack.

**Comment by jgw96** [comment]
Hello, thanks for opening an issue with us! I am going to be closing this one as a duplicate of https://github.com/driftyco/ionic/issues/7225. Thanks for using Ionic!

**Comment by dkobia** [comment]
@jgw96 so you don't think that its a problem that `dismiss()` is fired twice? Once when dismiss() is actually called and second when there is the defined duration is reached. If I call dismiss(), I don't think that when the duration is reached it should fire again. I don't think this is a Promise issue.

**Comment by flash716** [comment]
@dkobia I agree. I've attempted to use the promised returned by the NavController as defined in #7225  and still always get multiple onDismiss events. Currently I've worked around this by using the promise `loading.dismiss().then()`. It's not how the docs say to do this, but it currently works for my needs. On a side note when looking for the Loading docs I noticed that they have changed to an injectable LoadingController. It's not in the beta.10 release yet, so it looks like I'll need to revisit this in beta.11 anyways.

plnkr of my workaround: http://plnkr.co/edit/PXiRC0nBXyjTz9Sc0kap?p=preview

**Comment by jgw96** [comment]
Ahh @dkobia sorry, this was not an exact duplicate of the other issue. I am going to reopen this issue and look into why the dismiss triggered by the duration still gets fired after the component has already been dismissed.

**Comment by jgw96** [comment]
Hey @dkobia and @flash716 ! I just fixed this and it will be in beta.11. Now if you dismiss a loading component before the duration is done it will go ahead and clear the duration on the component so that dismiss does not fire twice.

**Comment by jgw96** [comment]
Also @flash716 you are correct that the loading component, along with all the overlay components have been refactored for beta.11 and should work much better for you in that release! They will now work more like native alerts etc. Thanks for using Ionic!

**Comment by dkobia** [comment]
Awesome @jgw96 -- thanks!

**Comment by jgw96** [comment]
@dkobia no problem 😄

**Comment by InfinityOne** [comment]
@jgw96 is this fixes for toasts too?

**Comment by jgw96** [comment]
@InfinityOne actually that issue should be solved by the overlay components refactor we did for beta.11. Thanks for using Ionic!

**Comment by jgw96** [comment]
This has actually already been fixed for the toast component too. Thanks for using Ionic!

**Comment by nickwinger** [comment]
this is the most annoying bug right now, the app is not useable with this. every modal that opens in a few seconds after the app started (e.g. login, passwortReset, confirmEmail etc.) gets closed...

if this has been fixed, when will beta11 be released ?

i cannot go on working right now with this bug... is there a workaround at least ??

**Comment by dkobia** [comment]
@nickwinger yes beta 11 has quite a few fixes to some really annoying issues from beta 10. Beta 11 adds even more disruptive changes, necessary [maybe], but disruptive cc @jgw96

For now @nickwinger, the workaround is to leave out `duration: xxx` from your Loading component

**Comment by nickwinger** [comment]
@dkobia thank you very much ! That really solved the issue for me.
Actually i also don't need the duration in my code, i had just copy&pasted it from somewhere.

**Comment by jgw96** [comment]
@dkobia @nickwinger There are a good amount of changes to overlay components like modals, loading etc in beta.11, but its definitely for the better! They now work more like native overlay components and so far with our testing the refactor has fixed a lot of issues. We always do an upgrade guide when we release a new build, so there will be clear instructions on how to update when we release beta.11, and all the docs will be updated too. Hope this clears everything up!

**Comment by dkobia** [comment]
@jgw96 just a heads up that this issue persists even with your changes. Two dismiss events still occur if `duration` is specified.

**Comment by TiagoSilvaPereira** [comment]
I have the same problem now, a need to remove the ```duration``` from my Loading component.
